targets/csharp/source/PlayFabSDK/source/Qos/PlayFabQosApi.cs

```diff
@@ -0,0 +1,163 @@
+                return;
+            }
+
+            foreach (QosServer qosServer in response.Result.QosServers)
+            {
+                if (!qosServer.Region.HasValue)
+                {
+                    continue;
+                }
+
+                _dataCenterMap[qosServer.Region.Value] = qosServer.ServerUrl;
+            }
+        }
+
+        private async Task<List<QosRegionResult>> GetSortedRegionLatencies(int timeoutMs)
+        {
+            List<Task<QosRegionResult>> asyncPingResults = _dataCenterMap.Select(async kvp =>
+            {
+                var regionPinger = new RegionPinger(kvp.Value, kvp.Key);
+                return await regionPinger.PingAsync(timeoutMs);
+            }).ToList();
+
+            await Task.WhenAll(asyncPingResults);
+            List<QosRegionResult> results = new List<QosRegionResult>(asyncPingResults.Count);
+            foreach (Task<QosRegionResult> asyncPingResult in asyncPingResults)
+            {
+                results.Add(asyncPingResult.Result);
+            }
+
+            results.Sort((x,y) => x.LatencyMs.CompareTo(y.LatencyMs));
+
+            return results;
+        }
+
+        private async Task SendResultsToPlayFab(QosResult result)
+        {
+            var eventContents = new EventContents
+            {
+                Name = "qos_result",
+                EventNamespace = "playfab.servers",
+                Payload = result
+            };
+
+            var writeEventsRequest = new WriteEventsRequest
+            {
+                Events = new List<EventContents> {eventContents}
+            };
+
+            PlayFabResult<WriteEventsResponse> response = null;
+            try
+            {
+                response = await WriteTelemetryAsync(writeEventsRequest);
+            }
+            catch (Exception exception)
+            {
+                LogQos(exception.ToString());
+            }
+
+            if (response == null || response.Error != null)
+            {
+                LogQos(response?.Error.GenerateErrorReport());
+            }
+        }
+
+        private static async Task<PlayFabResult<WriteEventsResponse>> WriteTelemetryAsync(WriteEventsRequest request, object customData = null, Dictionary<string, string> extraHeaders = null)
+        {
+            if ((request?.AuthenticationContext?.EntityToken ?? PlayFabSettings.staticPlayer.EntityToken) == null) throw new PlayFabException(PlayFabExceptionCode.EntityTokenNotSet, "Must call GetEntityToken before calling this method");
+
+            var httpResult = await PlayFabHttp.DoPost("/Event/WriteTelemetryEvents", request, "X-EntityToken", PlayFabSettings.staticPlayer.EntityToken, extraHeaders);
+            if (httpResult is PlayFabError)
+            {
+                var error = (PlayFabError)httpResult;
+                PlayFabSettings.GlobalErrorHandler?.Invoke(error);
```

Comment:
Please remove this. There's a high chance we will remove GlobalErrorHandler later.